### PR TITLE
[JENKINS-46602] Java 9 support: requires bumping asm to 6

### PR DESCRIPTION
**Pull request by batmat**
To support Java 9+, we need to bump to ASM 6.

@reviewbybees

### CODE REVIEW

**Comment by daniel-beck** [comment]
Specifically, https://issues.jenkins-ci.org/browse/JENKINS-46602, no?

**Comment by batmat** [comment]
@daniel-beck yes, I think it's actually a global issue, but right for now, adjusting.

**Comment by batmat** [comment]
Added @jtnord @jglick & @kohsuke as reviewers as they were last committers or mentioned in the commit log.

**Comment by batmat** [comment]
Added @jtnord @jglick & @kohsuke as reviewers as they were last committers or mentioned in the commit log.

**Comment by jtnord** [comment]
I guess the classfile is rewritten correctly for https://bugs.openjdk.java.net/browse/JDK-8148785 (esp around string concatenation).

IIRC we write the classfile with the highest of JDK7 or the original version, so we are not going from pre JDK9 to JDK9 in any given class so should be ok AFAICT.

**Comment by daniel-beck** [comment]
Could we get a core PR applying this change with a snapshot? Seems useful even outside JRE 9 compatibility.

File: pom.xml [hunk 1]
```diff
@@ -114,8 +114,8 @@
     </dependency>
     <dependency>
       <groupId>org.kohsuke</groupId>
-      <artifactId>asm5</artifactId>
-      <version>5.0.1</version>
+      <artifactId>asm6</artifactId>
```
Comment (jglick) [inline pom.xml]: Downstream of something in our repackaging?

Comment (batmat) [reply pom.xml]: Yes @jglick https://github.com/kohsuke/package-renamed-asm/pull/3 released since by Kohsuke.

Comment (batmat) [reply pom.xml]: Yes @jglick https://github.com/kohsuke/package-renamed-asm/pull/3 released since by Kohsuke.

**Comment by batmat** [comment]
@daniel-beck good point. Done as https://github.com/jenkinsci/jenkins/pull/3018

**Comment by jglick** [comment]
Some FindBugs failures it seems.

**Comment by jglick** [comment]
Note that the `master` version depends on `org.kohsuke:asm5:5.0.1`. I am not sure what wacky Gradle mistakes are to blame, but this not only includes the repackaged ASM, it also includes ASM (5, I suppose) under the _original_ package names, where it will of course clash with other stuff. The current `package-renamed-asm/pom.xml` in the `asm5` branch builds fine, as I presume 5.0.4 did, but…this release was never staged properly and is nowhere to be found. Sigh.

**Comment by jglick** [comment]
> it also includes ASM […] under the _original_ package names

Actually not true—it includes empty ZIP directory entries for the original package names, but no actual class files.

**Comment by batmat** [comment]
Yes, FindBugs is unrelated (pre-failing). Filed a fix separately on https://github.com/jenkinsci/bytecode-compatibility-transformer/pull/10

Thanks!

**Comment by batmat** [comment]
Merged with #10, will merge after/if a successful PR build.

**Comment by batmat** [comment]
FTR, this got released in `bytecode-compatibility-transformer-2.0-alpha-3`

**Pull request merged**